### DISCUSSION TITLE
Meta-predicates

### DISCUSSION THREAD

**Discussion by weitkaemper** [LogtalkDotOrg/logtalk3#230] [Q&A]
I am currently porting a statistical relational machine learning system from plain Prolog to Logtalk, and I have managed to reproduce the original system's behaviour on SWI in normal compilation mode. However, when loading with the optimized flag to "on", Logtalk throws existence errors on queries whose behaviour was correct in normal mode.
It seems to have to do with the resolution of meta-predicates, and the calling contexts of goals within them. 

My questions:
Are there known discrepancies here between optimized and normal compilation mode?
Does anyone have any experience as to how best to debug such errors, given that they don't occur in any debug mode?

**Comment by pmoura** [comment]
There should be no differences (other than performance). Can you post a reproducible example and the errors you get? Is there a public repo with the code?

Wondering if you're calling SWI-Prolog library meta-predicates (if so, take a look at the Logtalk [`meta`](https://logtalk.org/library/library_index.html#meta) and [`meta_compiler`](https://logtalk.org/library/library_index.html#meta-compiler) libraries) and there's some issue with SWI-Prolog auto-loading mechanism.

**Comment by pmoura** [comment]
Possibly related: https://github.com/LogtalkDotOrg/logtalk3/discussions/143

**Comment by weitkaemper** [comment]
Thank you for your very quick replies! 
I included the source files of a small component to illustrate the issue. The call
```
kb::compute_gradients(father(A, B), [[0.002, 0.6, -0.01]:::father(A, B)<=((true, childof(B, C)), childof(D, A))], G).
```
should compute a list of different numbers, some positive, some negative. 
This works in normal compilation mode. In optimised mode, however, it returns a list of only 0.5 and -0.5 (which would be the case exactly if there were empty background knowledge). 

[logtalk_081124_mwe.zip](https://github.com/user-attachments/files/17683477/logtalk_081124_mwe.zip)

**Reply by weitkaemper** [reply]
As so often, I have seen the mistake just after I posted it (the meta-predicate declaration of aggregate:aggregate_all/3 in loader.lgt was incorrect).  So the only remaining question (if that is still interesting) would be why that was harmless in normal mode but not in optimize mode (and why it still doesn't work in my full condebase).

**Reply by pmoura** [reply]
The code optimizations enabled by using the `optimize(on)` compiler option require accurate meta-predicate template information. In normal mode, in this particular case, the meta-predicate call is being resolved at runtime.

Btw, and as you asked about debugging, are you already using the VSCode Logtalk extension?

https://marketplace.visualstudio.com/items?itemName=LogtalkDotOrg.logtalk-for-vscode

See its readme for details on debugging support.

**Reply by weitkaemper** [reply]
Thank you very much for the pointer! Unfortunately I am an incorrigible emacs user, so I  very much appreciate the provided emacs mode instead :)

**Reply by pmoura** [reply]
https://marketplace.visualstudio.com/items?itemName=tuttieee.emacs-mcx

**Comment by pmoura** [comment]
I found why you get different results between compiling the code in normal mode vs compiling the code in optimal mode. Your meta-predicate declarations are incorrect as the meta-argument in the `compute_gradients/3` predicate is not a goal but a term that *contains* goals (see the Logtalk Handbook [section on meta-predicates](https://logtalk.org/manuals/userman/predicates.html#meta-predicate-directive) for details). Use instead:

```logtalk
:- meta_predicate(compute_gradients(*,::,*)).
```

Also, you don't need the `meta_predicate/1` directive for the `aggregate_all/3` predicate. In your `loader.lgt` file, simply load the `aggregate ` library using:

```logtalk
:- use_module(library(aggregate),[]).
```

**Reply by weitkaemper** [reply]
Thank you, I have corrected the declaration on compute_gradients and adapted the declaration in kb as suggested in the thread you linked earlier. 
I am probably misuderstanding how the use_module/2 declarations work in Logtalk. I mainly program in XSB, and one of several reasons for the Lotalk port is to bridge the gap between the module systems of  SWI-Prolog and XSB for making my XSB program available on SWI without losing XSB compatibility. In XSB, all module imports must specify the exact predicates you are loading. However, the naive port (replacing library(aggregate) with c_aggregate) does not run on XSB, with the error 
```Existence (No procedure calculate_groundings/3 exists)```
which again suggests issues with the calling context. 
I am attaching the updated zip file which could now work with XSB but doesn't; any pointers are appreciated!
[logtalk_081124_mwe.zip](https://github.com/user-attachments/files/17686610/logtalk_081124_mwe.zip)

**Reply by pmoura** [reply]
Simply move the following directive from the `loader.lgt` file to inside the `scorer` object:

```logtalk
:- meta_predicate(c_aggregate:aggregate_all(*,0,*)).
```

**Reply by pmoura** [reply]
You can also make the code run on both SWI-Prolog and XSB easily. In the `loader.lgt` file:

```logtalk
:- if(current_logtalk_flag(prolog_dialect, swi)).
	:- use_module(library(aggregate), []).
:- elif(current_logtalk_flag(prolog_dialect, xsb)).
	:- ensure_loaded(c_aggregate).
:- endif.
```

Inside the `scorer` object:

```logtalk
:- if(current_logtalk_flag(prolog_dialect, swi)).
	:- use_module(aggregate, [aggregate_all/3]).
:- elif(current_logtalk_flag(prolog_dialect, xsb)).
	:- use_module(c_aggregate, [aggregate_all/3]).
	:- meta_predicate(c_aggregate:aggregate_all(*,0,*)).
:- endif.
```

Also, call the `aggregate_all/3` using implicit qualification.

**Reply by weitkaemper** [reply]
That worked brilliantly, thank you. Is it a general rule that meta-predicate directives for imported predicates should be included in every object that calls them, rather than globally in the loader file?

**Reply by pmoura** [reply]
Depends on the Prolog backend and the `meta_predicate/1` directives (if present). A consequence of the lack of Prolog standardization. In general, you don't need to provide or override `meta_predicate/1` directives for module meta-predicates as long as these are unambiguous and can be accessed using `predicate_property/2`. In this particular case, for XSB, we have:

```text
| ?- predicate_property(c_aggregate:aggregate_all(_,_,_), P).

P = static;

P = exported;

no
```

Thus we need to provide ourselves the `meta_predicate/1` directive to enable Logtalk to properly compile calls to the meta-predicate. But this is not an issue with SWI-Prolog:

```text
?- predicate_property(aggregate:aggregate_all(_,_,_), P).
P = interpreted ;
P = visible ;
P = exported ;
P = static ;
P = transparent ;
P = (meta_predicate aggregate_all(?, 0, -)) ;
...
```

**Reply by weitkaemper** [reply]
> You can also make the code run on both SWI-Prolog and XSB easily. In the `loader.lgt` file:
> 
...
> 
> Also, call the `aggregate_all/3` using implicit qualification.

That's exactly what I am doing in my actual application. In fact, there is a lot more conditional compilation happening...
I just stripped that down to provide a sort-of minimal working example for this thread.

**Comment by pmoura** [comment]
The code you shared suggests an application of the *many worlds* design pattern. See e.g. https://logtalk.org/2019/11/13/many-worlds-design-pattern.html

**Reply by weitkaemper** [reply]
Thank you for the link! Inspired by the Metagol port, I had originally tried using inheritance for this, and I somehow got muddled on the right meta-predicate directives, and on the fact that the concrete predicates in the background knowledge are not known to me writing the scorer. So I thought it would be easier to have a service object that provides scoring as a meta-predicate. 
The actual Metagol architecture is not that easy to copy since I will eventually want to learn one classifier from many different knowledge bases, and so I don't want the entire learning algorithm incorporated in a single kb.

**Reply by pmoura** [reply]
Not sure I understand. An object representing a knowledge base can inherit from multiple learning algorithm objects (or import multiple learning algorithm categories).

**Reply by weitkaemper** [reply]
What I meant was that the central algorithm (not included in the mwe I posted above) eventually needs access to all the different knowledge bases, not just a single one. 
Therefore I would presumably want the central algorithm in a separate object, which then refers to the scorer(s) of the individual kb's. 
Just in case you are interested, I have a preprint discussing the prior state of the (pure Prolog) program here:
https://www.informatik.uni-wuerzburg.de/fileadmin/10030600/2024/KI_2004_paper_174.pdf

**Reply by weitkaemper** [reply]
In any case, thank you very much for helping me to get everything to work so quickly! 
As I implemented it, there is a ca. 100% runtime overhead on XSB (40% on SWI-Prolog) compared to the plain Prolog version, so I'll still be looking into that. Using include directives instead of a separate scorer entity speeds things up, but is probably not in the spirit of object-oriented programming...

**Reply by pmoura** [reply]
One problem hurting performance is the [circular dependency](https://logtalk3.readthedocs.io/en/latest/userman/performance.html#circular-references) between the `kb` and `scorer` objects. In the `loader.lgt` file, you're loading `scorer` before `kb`. That means that the compilation of the following directive in `scorer`:

```logtalk
:- uses(kb,[positive/1,negative/1,domain/1]).
```

uses message-sending instead of static binding for the calls to the `kb` predicates.

**Reply by pmoura** [reply]
> In any case, thank you very much for helping me to get everything to work so quickly! As I implemented it, there is a ca. 100% runtime overhead on XSB (40% on SWI-Prolog) compared to the plain Prolog version, so I'll still be looking into that. Using include directives instead of a separate scorer entity speeds things up, but is probably not in the spirit of object-oriented programming...

Expected Logtalk overhead compared with plain Prolog is in the low single digits. There's a Handbook [performance](https://logtalk.org/manuals/userman/performance.html) section. See also the [`benchmarks`](examples/benchmarks) example. I just run it with latest  versions of both SWI-Prolog and XSB. Got the expected results; no performance regressions.

Other than the issue I mentioned above with circular dependencies, I will need to look into the full code and how you're benchmarking to help diagnose the numbers you report. Is there a public repo?

**Reply by weitkaemper** [reply]
Sorry for the late reply! I am currently travelling and therefore also didn't have the chance to look into your comment below (although it sounds perfectly correct - I recently refactored this part of the code and clearly did not choose the most elegant formulation...). 
Unfortunately, the university gitlab repository where the current Prolog implementation is hosted recently needs special permissions  to make a repository public, so I am just attaching zip archives of both the pure Prolog version and my LogTalk port.
The entrypoint is boostrlr.pl / .lgt and the test commands are ```b_rlr(father(_,_),WC).``` and ```boostrlr::b_rlr(father(_,_),WC).``` respectively. 
[boostrlr_prolog.zip](https://github.com/user-attachments/files/17799182/boostrlr_prolog.zip)
[BoostRLR_LogTalk.zip](https://github.com/user-attachments/files/17799197/BoostRLR_LogTalk.zip)

PS Thank you very much for investing so much of your time into this! I am planning on publishing on this line of work at an ICLP or similar venue, so if you would be interested in any more formal collaboration, feel free to reach out! :)

**Reply by pmoura** [reply]
I'm on macOS and thus the `.so` libraries in those archives are not usable for me. I suggest adding a `Makefile` file for compiling the C code and that allows for different SWI-Prolog installation prefixes.

**Reply by pmoura** [reply]
However, I'm able to load the code in a Ubuntu 20 image (running on VirtualBox). With the sample queries you mention above, I don't observe the performance issues you mentioned. Using SWI-Prolog 9.3.14, the `b_rlr(father(_,_),WC)` goal took 21.151 seconds and the `boostrlr::b_rlr(father(_,_),WC)` goal took 22.571 seconds (loading `kb` before `scorer` lowers the time slightly.

**Reply by weitkaemper** [reply]
Which compiler flags have you used for your comparison? For SWI-Prolog, I compared performance of plain SWI-Prolog with the -O compiler flag with the optimized LogTalk version. I assumed that the optimized LogTalk flag would also set the optimization flag in SWI-Prolog.
XSB doesn't have a dedicated optimisation flag as far as I know, so I just compiled the code normally in plain XSB and optimized in LogTalk.

**Reply by pmoura** [reply]
Running the plain Prolog query using `swipl -O` takes 19.206 seconds, i.e. ~3 seconds faster than the Logtalk version. Logtalk sets by default the SWI-Prolog `optimise` flag to `true`. Given the performance issue I mentioned above with circular references, I find the small performance expected and reasonable.

**Reply by weitkaemper** [reply]
That is indeed completely reasonable, and I am not sure why the difference is somewhat larger for me (but still OK). Since I am personally using and running the program under XSB, I was more concerned with the larger performance drop there than under SWI-Prolog.

**Reply by pmoura** [reply]
Hard to tell. Trying the Prolog version with XSB in my Ubuntu 20 VM gives an error:

```text
| ?- time(b_rlr(father(_,_),WC)).
[Compiling /home/pmoura/xsbgit/XSB/prolog-commons/c_numlists]
[Module c_numlists compiled, cpu time used: 0.000 seconds]
% 0.0 CPU in  0.003 seconds (0% CPU)
++Error[XSB/Runtime/P]: [Existence (No predicate c_numlists : sum_list / 2 exists)] []
...
```

The Logtalk version is faster than SWI-Prolog:

```text
| ?- time(boostrlr::b_rlr(father(_,_),WC)).
% 12.201 CPU in  12.412 seconds (98% CPU)
...
```

Anyway, see the warning in the https://logtalk3.readthedocs.io/en/latest/userman/performance.html page regarding benchmarking at the top-level. Not an issue with SWI-Prolog as this system provides the necessary hooks to bypass the issue.

Also, the Logtalk (not LogTalk!) code you sent me requires hacking when switching between SWI-Prolog and XSB. First step is to not use any explicit module-qualification in the code but instead rely on `use_module/2` and `import/2` directives inside conditional compilation blocks. You'll  need to write the `import/2` directives in canonical form to avoid syntax errors when running the code with SWI-Prolog, however. For example, instead of writing:

```logtalk
:- import aggregate_all/3 from c_aggregate.
```

use instead:

```logtalk
:- import(from(/(aggregate_all,3),c_aggregate)).
```

**Reply by weitkaemper** [reply]
Thank you very much for the analysis, and my apologies for the delayed reply. I cannot currently reproduce the plain XSB error, but your reported performance seems more reasonable than mine. Is there a canonical way to measure performance while bypassing the top-level issues? Would wrapping the testing command in a static tester object be sufficient?

**Reply by pmoura** [reply]
I'm using the current XSB git version. Maybe some bug introduced after the 5.0.0 released. I already reported it to Theresa and David.

Indeed, wrapping the benchmark query in a static object compiled optimized would bypass top-level interpreter overheads. But if the query requires a non-trivial number of computations, the overheads should be small. As I mention above when discussing circular references, changing the files loader order may help (notably, `kb` before `scorer`).

**Comment by pmoura** [comment]
Btw, why:

```logtalk
compute_gradients(Head,WeightedClauses,Gradients) :-
    findall(Gradient,(
                term::varnumbers((Head,WeightedClauses),(ReHead,ReWeightedClauses)),
                compute_gradients_aux(ReHead,ReWeightedClauses,Gradient)),
            Gradients).
```

instead of:

```logtalk
compute_gradients(Head,WeightedClauses,Gradients) :-
    term::varnumbers((Head,WeightedClauses),(ReHead,ReWeightedClauses)),
    findall(Gradient, compute_gradients_aux(ReHead,ReWeightedClauses,Gradient)), Gradients).
```
